pyne/endl.py

```diff
@@ -0,0 +1,341 @@
+                low_xs = np.e ** interp(np.log(low))
+                xs = np.insert(xs[e_int > low], 0, low_xs)
+                e_int = np.insert(e_int[e_int > low], 0, low)
+            if high in e_int:
+                xs = xs[e_int <= high]
+                e_int = e_int[e_int <= high]
+            elif high is not None:
+                high_xs = np.e ** interp(np.log(high))
+                xs = np.append(xs[e_int < high], high_xs)
+                e_int = np.append(e_int[e_int < high], high)
+
+        de_int = float(e_int[-1]-e_int[0])
+        x1 = e_int[:-1]
+        x2 = e_int[1:]
+        y1 = xs[:-1]
+        y2 = xs[1:]
+        A = - np.log(y2/y1)/np.log(x1/x2)
+        B = - (np.log(y1)*np.log(x2) - np.log(y2)*np.log(x1))/np.log(x1/x2)
+        return np.nansum(np.e**B / (A+1) * (x2**(A+1) - x1**(A+1))/de_int)
+
+    def get_rx(self, nuc, p_in, rdesc, rprop, x1=None, p_out=None):
+        """get_rx(nuc, p_in, rdesc, rprop, x1=None, p_out=None)
+        Grab the data for one reaction type.
+
+        Parameters
+        ----------
+        nuc : int
+            id form of nucleus to read from.
+        p_in : int
+            ENDL incident particle designator
+        rdesc : int
+            ENDL reaction descriptor
+        rprop : int
+            ENDL reaction property
+        x1 : int or None
+            ENDL atomic subshell indicator (if applicable)
+        yo : int or None
+            ENDL outgoing particle designator (if applicable)
+
+        Returns
+        -------
+        data : NumPy array
+            Contains the reaction data in a n-by-m-dimensional array. The
+            values of n and m depend on the reaction property rprop.
+        """
+        nuc = nucname.id(nuc)
+        if nuc in self.structure:
+            return self._read_nuc_pin_rdesc_rprop(nuc, p_in,
+                                                  rdesc, rprop,
+                                                  x1, p_out)
+        else:
+            raise ValueError('Nucleus {} does not exist.'.format(nuc))
+
+    def _read_nuc_pin_rdesc_rprop(self, nuc, p_in, rdesc, rprop, x1, yo):
+        """Load in the data from one reaction into self.structure.
+
+        Parameters
+        ----------
+        nuc : int
+            id of nuclide.
+        p_in : int
+            ENDL incident particle designator
+        rdesc : int
+            ENDL reaction descriptor
+        rprop : int
+            ENDL reaction property
+        x1 : int or None
+            ENDL atomic subshell indicator (if applicable)
+        yo : int or None
+            ENDL outgoing particle designator (if applicable)
+
+        Returns
+        -------
+        array, float64
+            float64 NumPy array containing the reaction data. The shape of the
+            array depends on the ENDL reaction property.
+        """
+        opened_here = False
+
+        try:
+            pdp_dict = self.structure[nuc]['pin_rdesc_rprop']
+        except KeyError as e:
```

Comment:
So I always interpreted that to mean that 

```python
pdp_dict = self.structure[nuc]['pin_rdesc_rprop'] 
```

on its on is more Pythonic (which it is), and if an error occurs you catch it somewhere farther up in the stack. Also the EAFP comparison is a little disingenuous since C exception handling syntax. However, if you are going to test for inclusion at the call site (though exceptions or conditionals), I think that there is an argument to be made for using conditionals based on their timings.  Consider the following:

```python
scopatz@athenaie ~ $ def f(test):
`·.,¸,.·*¯`·.,¸,.·*¯     if test:
`·.,¸,.·*¯`·.,¸,.·*¯         pass
`·.,¸,.·*¯`·.,¸,.·*¯     else:
`·.,¸,.·*¯`·.,¸,.·*¯         pass
`·.,¸,.·*¯`·.,¸,.·*¯ 
scopatz@athenaie ~ $ def g(test):
`·.,¸,.·*¯`·.,¸,.·*¯     try:
`·.,¸,.·*¯`·.,¸,.·*¯         assert test
`·.,¸,.·*¯`·.,¸,.·*¯     except AssertionError:
`·.,¸,.·*¯`·.,¸,.·*¯         pass
`·.,¸,.·*¯`·.,¸,.·*¯ 
scopatz@athenaie ~ $ timeit "f(True)"
The slowest run took 130.28 times longer than the fastest. This could mean that an intermediate result is being cached.
10000000 loops, best of 3: 76.8 ns per loop
scopatz@athenaie ~ $ timeit "f(False)"
10000000 loops, best of 3: 75.8 ns per loop
scopatz@athenaie ~ $ timeit "g(True)"
10000000 loops, best of 3: 91.4 ns per loop
scopatz@athenaie ~ $ timeit "g(False)"
The slowest run took 59.51 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 387 ns per loop
```

The False case for exceptions is *much* slower.